Assert hit in WebKit::RemoteImageBuffer::getShareableBitmap

NKRosario opened the pull request:
#### 163fdbfc157be7bccec29ee8cf942b09c726e77c
<pre>
Assert hit in WebKit::RemoteImageBuffer::getShareableBitmap
<a href="https://bugs.webkit.org/show_bug.cgi?id=273285">https://bugs.webkit.org/show_bug.cgi?id=273285</a>
<a href="https://rdar.apple.com/122378565">rdar://122378565</a>

Reviewed by Alex Christensen.

The `IPC.messages.RemoteRenderingBackend_CreateImageBuffer` was able to send PixelFormat
values that are not valid and therefore caused a crash. Added an enum for only the valid
PixelFormat values so that only valid values can be sent over IPC

* LayoutTests/ipc/send-gpu-GetShareableBitmap-RemoteRenderingBackend-expected.txt: Added.
* LayoutTests/ipc/send-gpu-GetShareableBitmap-RemoteRenderingBackend.html: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::backendParameters):
* Source/WebCore/platform/graphics/ImageBuffer.h:
(WebCore::ImageBuffer::create):
(WebCore::ImageBuffer::pixelFormatValidated const):
(WebCore::ImageBuffer::pixelFormat const): Deleted.
* Source/WebCore/platform/graphics/ImageBufferPixelFormat.h: Added.
(WebCore::convertToImageBufferPixelFormat):
(WebCore::convertToPixelFormat):
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
(WebKit::RemoteImageBuffer::getShareableBitmap):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::isSmallLayerBacking):
(WebKit::allocateImageBufferInternal):
(WebKit::RemoteRenderingBackend::createImageBuffer):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::sinkIntoBufferForDifferentThread):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
(WebKit::RemoteImageBufferProxy::create):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::createRemoteImageBuffer):
(WebKit::RemoteRenderingBackendProxy::createDisplayListRecorder):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Tools/TestWebKitAPI/Tests/WebCore/DisplayListRecorderTests.cpp:
(TestWebKitAPI::createTestImageBuffer):

Canonical link: <a href="https://commits.webkit.org/279951@main">https://commits.webkit.org/279951@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04dc9bcb8200e3fc6f14795c998eafe7f37d45e5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54943 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34396 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7532 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58221 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5674 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57243 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41944 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5705 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44496 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3846 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57038 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32482 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47597 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25622 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29271 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4939 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3815 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/51128 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5162 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59811 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30207 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5317 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51914 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31339 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47674 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51349 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12101 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32354 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31128 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
```diff
@@ -44,6 +44,7 @@
 #include "StreamClientConnection.h"
 #include "ThreadSafeObjectHeap.h"
 #include "WorkQueueMessageReceiver.h"
+#include <WebCore/ImageBufferPixelFormat.h>
```
I think this header should be in the source RemoteRenderingBackendProxy.cpp.

shallawa commented on Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
```diff
@@ -664,10 +664,13 @@ void GPUConnectionToWebProcess::createRenderingBackend(RemoteRenderingBackendCre
         return IPC::ScopedActiveMessageReceiveQueue { RemoteRenderingBackend::create(*this, WTFMove(creationParameters), streamConnection.releaseNonNull()) };
     });
-    ASSERT_UNUSED(addResult, addResult.isNewEntry);
+    if (!addResult.isNewEntry) {
+        streamConnection->invalidate();
+        MESSAGE_CHECK(false);
+    }
```
I think this change is not related to the PixelFormat change. Can it be in a separate PR?

NKRosario replied:
it's needed for the IPC messages in the test

shallawa commented on Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
```diff
@@ -120,6 +120,8 @@ void RemoteImageBuffer::getShareableBitmap(WebCore::PreserveResolution preserveR
         auto backendSize = m_imageBuffer->backendSize();
         auto logicalSize = m_imageBuffer->logicalSize();
         auto resultSize = preserveResolution == WebCore::PreserveResolution::Yes ? backendSize : m_imageBuffer->truncatedLogicalSize();
+        if (resultSize.isEmpty())
+            return std::nullopt;
```
How is related to this PR?

NKRosario replied:
I was hitting an assert without it so I added it so the fuzzer can still pass through

shallawa commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
```diff
@@ -302,10 +302,10 @@ RefPtr<ImageBuffer> RemoteRenderingBackend::allocateImageBuffer(const FloatSize&
 {
     assertIsCurrent(workQueue());
-    RefPtr<ImageBuffer> imageBuffer = allocateImageBuffer(logicalSize, renderingMode, purpose, resolutionScale, colorSpace, pixelFormat, { }, imageBufferIdentifier);
+    RefPtr<ImageBuffer> imageBuffer = allocateImageBuffer(logicalSize, renderingMode, purpose, resolutionScale, colorSpace, convertToPixelFormat(imageBufferPixelFormat), { }, imageBufferIdentifier);
```
This is confusing. 

1. Here `convertToPixelFormat(imageBufferPixelFormat)` is called because `allocateImageBuffer()` takes `PixelFormat`
2. It passes it to `allocateImageBufferInternal()` which passes it to `ImageBuffer::create()`
3. `ImageBuffer::create()` calls `convertToImageBufferPixelFormat(pixelFormat)` because `ImageBufferParameters` now has `ImageBufferPixelFormat`

So we did the conversion twice. Why do not we make everything `ImageBufferPixelFormat`? Only when we want a PixelBuffer we can call `convertToPixelFormat(imageBufferPixelFormat)`.

shallawa commented on Source/WebCore/platform/graphics/ImageBufferPixelFormat.h:
```diff
@@ -0,0 +1,87 @@
+        ASSERT_NOT_REACHED();
+        return ImageBufferPixelFormat::BGRX8;
+#endif
+    }
```
```
   switch (format) {
    case PixelFormat::BGRX8:
        return ImageBufferPixelFormat::BGRX8;
    case PixelFormat::BGRA8:
        return ImageBufferPixelFormat::BGRA8;
#if HAVE(IOSURFACE_RGB10)
    case PixelFormat::RGB10:
        return ImageBufferPixelFormat::RGB10;
    case PixelFormat::RGB10A8:
        return ImageBufferPixelFormat::RGB10A8;
#endif
    default:
        break;
    }
```
This might be clearer and represents the opposite of the switch statement below.

shallawa commented on Source/WebCore/platform/graphics/ImageBufferPixelFormat.h:
```diff
@@ -0,0 +1,87 @@
+
+namespace WebCore {
+enum class PixelFormat : uint8_t;
+enum class ImageBufferPixelFormat : uint8_t {
```
Please add a blank line between the `namespace` and the forward declaration and a line between the forward declaration and the new enum.

shallawa commented on Source/WebCore/platform/graphics/ImageBuffer.h:
```diff
@@ -157,7 +158,8 @@ class ImageBuffer : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Image
     RenderingPurpose renderingPurpose() const { return m_parameters.purpose; }
-    PixelFormat pixelFormat() const { return m_parameters.pixelFormat; }
+    ImageBufferPixelFormat imageBufferPixelFormat() const { return m_parameters.imageBufferPixelFormat; }
+    PixelFormat pixelFormat() const { return convertToPixelFormat(m_parameters.imageBufferPixelFormat); }
```
Why do we need to expose both? I think most of the time the `ImageBufferPixelFormat` will be needed. Only when we get a `PixelBuffer` from the `ImageBuffer`, the `PixelFormat` will be needed.

shallawa commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -114,7 +114,7 @@ IntSize ImageBuffer::calculateBackendSize(FloatSize logicalSize, float resolutio
 ImageBufferBackendParameters ImageBuffer::backendParameters(const ImageBufferParameters& parameters)
 {
-    return { calculateBackendSize(parameters.logicalSize, parameters.resolutionScale), parameters.resolutionScale, parameters.colorSpace, parameters.pixelFormat, parameters.purpose };
+    return { calculateBackendSize(parameters.logicalSize, parameters.resolutionScale), parameters.resolutionScale, parameters.colorSpace, convertToPixelFormat(parameters.imageBufferPixelFormat), parameters.purpose };
```
This is confusing. Why `ImageBufferParameters` has `ImageBufferPixelFormat` but `ImageBufferBackendParameters` has `PixelFormat`? Can't we have `ImageBufferPixelFormat` in both?

shallawa commented on Source/WebCore/platform/graphics/ImageBufferPixelFormat.h:
```diff
@@ -0,0 +1,86 @@
+#endif
+};
+
+WEBCORE_EXPORT static constexpr ImageBufferPixelFormat convertToImageBufferPixelFormat(PixelFormat format)
```
I think this function should not be `WEBCORE_EXPORT static`. A `constexpr` function is implicitly inline.

shallawa commented on Source/WebCore/platform/graphics/ImageBufferPixelFormat.h:
```diff
@@ -0,0 +1,86 @@
+    return ImageBufferPixelFormat::BGRX8;
+}
+
+WEBCORE_EXPORT static constexpr PixelFormat convertToPixelFormat(ImageBufferPixelFormat format)
```
Ditto.

shallawa commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -79,19 +79,24 @@ RefPtr<ImageBuffer> ImageBuffer::create(const FloatSize& size, RenderingPurpose
+RefPtr<ImageBuffer> ImageBuffer::create(const FloatSize& size, RenderingPurpose purpose, float resolutionScale, const DestinationColorSpace& colorSpace, PixelFormat pixelFormat, OptionSet<ImageBufferOptions> options, GraphicsClient* graphicsClient)
+{
+    return RefPtr<ImageBuffer>(ImageBuffer::create(size, purpose, resolutionScale, colorSpace, convertToImageBufferPixelFormat(pixelFormat), options, graphicsClient));
 }
```
Why do we still need this?

NKRosario replied:
I added that because there were lots of files that used `create(const FloatSize&, RenderingPurpose, float resolutionScale, const DestinationColorSpace&, PixelFormat)` and it seemed cleaner to add that function with the conversion than adding a bunch of files to the PR but I can do that instead if that's better.

shallawa commented on Source/WebCore/platform/graphics/ImageBuffer.h:
```diff
@@ -85,19 +86,20 @@ struct ImageBufferParameters {
 public:
     using Parameters = ImageBufferParameters;
+    WEBCORE_EXPORT static RefPtr<ImageBuffer> create(const FloatSize&, RenderingPurpose, float resolutionScale, const DestinationColorSpace&, ImageBufferPixelFormat, OptionSet<ImageBufferOptions> = { }, GraphicsClient* graphicsClient = nullptr);
     WEBCORE_EXPORT static RefPtr<ImageBuffer> create(const FloatSize&, RenderingPurpose, float resolutionScale, const DestinationColorSpace&, PixelFormat, OptionSet<ImageBufferOptions> = { }, GraphicsClient* graphicsClient = nullptr);
```
I think there should be only one `create()` which replaces  `PixelFormat` with `ImageBufferPixelFormat`.

shallawa commented on Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
```diff
@@ -50,7 +50,7 @@ class RemoteImageBufferProxy : public WebCore::ImageBuffer {
     template<typename BackendType>
     static RefPtr<RemoteImageBufferProxy> create(const WebCore::FloatSize& size, float resolutionScale, const WebCore::DestinationColorSpace& colorSpace, WebCore::PixelFormat pixelFormat, WebCore::RenderingPurpose purpose, RemoteRenderingBackendProxy& remoteRenderingBackendProxy, bool avoidBackendSizeCheck = false)
```
Why is this function still taking `WebCore::PixelFormat`? Should not it take `WebCore::ImageBufferPixelFormat`?

NKRosario replied:
updated: going through the upstream functions I do see it converting from ImageBufferPixelFormat -> PixelFormat which will be converted back so I'll update this to use WebCore::ImageBufferPixelFormat

it seemed a lot cleaner to leave this as `WebCore::PixelFormat` and do the conversion than to change all the upstream functions, and it didn't have the convert and convert back issue that the IPC path had

kkinnunen-apple commented on Source/WebCore/platform/graphics/ImageBufferPixelFormat.h:
```diff
@@ -0,0 +1,86 @@
+#endif
+};
+
+constexpr ImageBufferPixelFormat convertToImageBufferPixelFormat(PixelFormat format)
```
If this function is needed at all, it should be returning `std::optional<ImageBufferPixelFormat>` and then the runtime release assert should be when the return is unexpectedly nullopt that cannot be handled..

kkinnunen-apple commented on Source/WebCore/platform/graphics/ImageBufferPixelFormat.h:
```diff
@@ -0,0 +1,86 @@
+#endif
+    }
+
+    ASSERT_NOT_REACHED();
```
These two lines aren't strictly needed, as the expectation is that the enum contains only the enum values.

kkinnunen-apple commented on Source/WebCore/platform/graphics/ImageBuffer.h:
```diff
@@ -85,19 +86,20 @@ struct ImageBufferParameters {
     float resolutionScale;
     DestinationColorSpace colorSpace;
-    PixelFormat pixelFormat;
+    ImageBufferPixelFormat imageBufferPixelFormat;
```
The variable name can still be pixelFormat and in all other places.

kkinnunen-apple commented on Source/WebCore/platform/graphics/ImageBuffer.h:
```diff
@@ -156,7 +158,7 @@ class ImageBuffer : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Image
     
     RenderingPurpose renderingPurpose() const { return m_parameters.purpose; }
-    PixelFormat pixelFormat() const { return m_parameters.pixelFormat; }
+    ImageBufferPixelFormat imageBufferPixelFormat() const { return m_parameters.imageBufferPixelFormat; }
```
The accessor should be just pixelFormat()

kkinnunen-apple reviewed (commented):
Thanks, looks good already. When you change more call sites to use ImageBufferPixelFormat, you probably see a lot of the changes disappear.

kkinnunen-apple commented on LayoutTests/ipc/send-gpu-GetShareableBitmap-RemoteRenderingBackend-expected.txt:
```diff
@@ -0,0 +1,5 @@
+layer at (0,0) size 800x600
```
The test should be marked as text test, and you should print out something like "DONE" so that it's clear that the test ran and did not crash. You can take a look at the other similar tests

shallawa commented on LayoutTests/ipc/send-gpu-GetShareableBitmap-RemoteRenderingBackend.html:
```diff
@@ -0,0 +1,51 @@
+        testRunner.waitUntilDone();
+        setTimeout(() => {
+            testRunner.notifyDone();
+        }, 10000);
```
10 seconds is a very long time for a test to take. Besides using `setTimeout()` still make it flaky especially for busy and slow machines on EWS. Can't the code below call `testRunner.notifyDone()` when the asynchronous waits receive the replies back form GPUP?

NKRosario replied:
`mac-wk2-stress` is timing out now when I switched to using `testRunner.notifyDone()` after  the IPC sends/waits

NKRosario replied:
nevermind I think it was just hitting the TypeError which isn't actually a problem. Checking that now

shallawa commented on Source/WebCore/html/HTMLVideoElement.cpp:
```diff
@@ -329,9 +329,10 @@ std::optional<DestinationColorSpace> HTMLVideoElement::colorSpace() const
+RefPtr<ImageBuffer> HTMLVideoElement::createBufferForPainting(const FloatSize& size, RenderingMode renderingMode, const DestinationColorSpace& colorSpace, ImageBufferPixelFormat pixelFormat) const
 {
     auto* hostWindow = document().view() && document().view()->root() ? document().view()->root()->hostWindow() : nullptr;
+
```
No need for this extra line.

shallawa commented on Source/WebCore/platform/graphics/ImageBufferPixelFormat.h:
```diff
@@ -0,0 +1,60 @@
+
+#pragma once
+
+#if HAVE(IOSURFACE)
+#include "IOSurface.h"
+#endif
```
I think this header file is not needed.

NKRosario replied:
wouldn't `#include "IOSurface.h"` be needed to move over the `formatForPixelFormat` function?

shallawa commented on Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
```diff
@@ -651,30 +651,27 @@ static IntSize computeMaximumSurfaceSize()
 }
 
-IOSurface::Format IOSurface::formatForPixelFormat(PixelFormat format)
+IOSurface::Format IOSurface::formatForPixelFormat(ImageBufferPixelFormat format)
```
This function is used by `ImageBufferBackend` only. Should not it be moved to ImageBufferPixelFormat.h:

```
constexpr PixelFormat convertToPixelFormat(ImageBufferPixelFormat format)
{
    ...
}

#if HAVE(IOSURFACE)
constexpr IOSurface::Format convertToIOSurfaceFormat(ImageBufferPixelFormat format)
{
    ...
}
#endif
```

shallawa commented on Tools/TestWebKitAPI/Tests/WebCore/DisplayListRecorderTests.cpp:
```diff
@@ -32,6 +32,7 @@
 #include <WebCore/DisplayListDrawingContext.h>
 #include <WebCore/GraphicsContext.h>
 #include <WebCore/ImageBuffer.h>
+#include <WebCore/ImageBufferPixelFormat.h>
```
No need to include this header because it is included in ImageBuffer.h.

shallawa reviewed (commented):
The change looks good. But I still do not understand what was wrong with [278192@main](https://commits.webkit.org/278192@main) and why it regressed the  nytimes rendering. More importantly I still do not understand what did you do differently in this PR to fix this regression.

kkinnunen-apple commented on LayoutTests/ipc/send-gpu-GetShareableBitmap-RemoteRenderingBackend.html:
```diff
@@ -0,0 +1,51 @@
+    try {
+        o35.sendSyncMessage(imageBufferID,IPC.messages.RemoteImageBuffer_GetShareableBitmap.name,0.1,[{type: 'bool',value: 0}]);
+    } catch { }
+    o35.invalidate();
```
So here you'd call notifyDone()
In case that is not always called, it means that something is wrong with the testcase. perhaps some exception or some such.

kkinnunen-apple reviewed (approved):
Looks good. Please still address the couple of said's comments:
- notifydone
- iosurface.h include that probably is not needed
- moving the enum conversion function to the imagebufferbackend that calls the function

shallawa commented on LayoutTests/ipc/send-gpu-GetShareableBitmap-RemoteRenderingBackend.html:
```diff
@@ -0,0 +1,96 @@
+        if (!(error instanceof TypeError)) {
+            console.log("Failed: " + error)
+            testRunner.notifyDone();
+            return
```
I think we do not need this return. The catch block will return anyway after exiting the if-statement.

shallawa commented on LayoutTests/ipc/send-gpu-GetShareableBitmap-RemoteRenderingBackend.html:
```diff
@@ -0,0 +1,96 @@
+    } catch (error) {
+        if (!(error instanceof TypeError)) {
+            console.log("Failed: " + error)
+            testRunner.notifyDone();
```
Why do we call `notifyDone()` every time we catch an exception? Should we continue to the next block and make the last statement call `notifyDone()`.

shallawa commented on LayoutTests/ipc/send-gpu-GetShareableBitmap-RemoteRenderingBackend.html:
```diff
@@ -0,0 +1,96 @@
+        o35.setSemaphores(resp[0].value, resp[1].value);
+    } catch (error) {
+        if (!(error instanceof TypeError)) {
+            console.log("Failed: " + error)
```
Can't we use js-test.js workflow here? The structure of the test can be the following:
```
<script src="../../resources/js-test.js"></script>
<script>
    description("description of the test");
    window.jsTestIsAsync = true;

    debug("description of the this call");
    try {
        ...
    } catch {
        if (!(error instanceof TypeError))
            debug("Failed: " + error);
    }
    // Other calls
    ...
    ...
    debug("description of the the last call");
    try {
        ...
    } catch {
        if (!(error instanceof TypeError))
            debug("Failed: " + error);
    } finally {
        finishJSTest();
    }
</script>
```

shallawa commented on LayoutTests/ipc/send-gpu-GetShareableBitmap-RemoteRenderingBackend.html:
```diff
@@ -0,0 +1,96 @@
+    } catch (error) {
+        if (!(error instanceof TypeError)) {
+            console.log("Failed: " + error)
+            testRunner.notifyDone();
```
These `notifyDone()` will end the test once an exception is thrown. There is no way to know how far the test goes before it ends. I think you need to log messages telling what exactly the script has tested.

shallawa commented on LayoutTests/ipc/send-gpu-GetShareableBitmap-RemoteRenderingBackend-expected.txt:
```diff
@@ -0,0 +1 @@
+This test passes if it does not crash.
```
This message is not sufficient anymore. The test can end once an exception is thrown. I think the test has to do all the steps and the expected file has to record that all the steps are tested.

NKRosario replied:
ok I figured it would fail if any of the `Failed: ` messages were also printed but I can put logs for each step

shallawa commented on Source/JavaScriptCore/build/XCBuildData/6783d0df4ec550e8a22a7ef8ca00cffe.xcbuilddata/build-request.json:
```diff
@@ -0,0 +1,27 @@
+{
```
What is this file doing?

NKRosario replied:
I didn't see that this file was added, will remove

shallawa commented on Source/JavaScriptCore/build/XCBuildData/6783d0df4ec550e8a22a7ef8ca00cffe.xcbuilddata/manifest.json:
```diff
@@ -0,0 +1 @@
+{"client":{"name":"basic","version":0,"file-system":"device-agnostic","perform-ownership-analysis":"no"},"targets":{"":["<all>"]},"commands":{"<all>":{"tool":"phony","inputs":["<WorkspaceHeaderMapVFSFilesWritten>"],"outputs":["<all>"]},"P0:::Gate WorkspaceHeaderMapVFSFilesWritten":{"tool":"phony","inputs":[],"outputs":["<WorkspaceHeaderMapVFSFilesWritten>"]}}}
```
What is the purpose of adding all these JavaScriptCore files?

NKRosario replied:
I didn't see that this was added, will remove

shallawa commented on LayoutTests/ipc/send-gpu-GetShareableBitmap-RemoteRenderingBackend.html:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    if (!window.IPC) {
+        testRunner.notifyDone();
```
Call `finishJSTest();` instead of calling `testRunner.notifyDone();`

shallawa commented on LayoutTests/ipc/send-gpu-GetShareableBitmap-RemoteRenderingBackend.html:
```diff
@@ -0,0 +1,90 @@
+            debug("Failed: " + error)
+    } finally {
+        o35.invalidate();
+        testRunner.notifyDone();
```
Call `finishJSTest();`

shallawa commented on LayoutTests/ipc/send-gpu-GetShareableBitmap-RemoteRenderingBackend.html:
```diff
@@ -0,0 +1,90 @@
+        testRunner.waitUntilDone();
+        testRunner.dumpAsText();
+        finishJSTest();
+    }
```
You do not need this. `window.jsTestIsAsync = true;` is exactly the same as `testRunner.waitUntilDone();`. Including `js-test.js` will call `testRunner.dumpAsText();`. Also you do not need to check `window.testRunner`.

NKRosario committed:
Assert hit in WebKit::RemoteImageBuffer::getShareableBitmap
https://bugs.webkit.org/show_bug.cgi?id=273285
rdar://122378565

Reviewed by Alex Christensen.

The `IPC.messages.RemoteRenderingBackend_CreateImageBuffer` was able to send PixelFormat
values that are not valid and therefore caused a crash. Added an enum for only the valid
PixelFormat values so that only valid values can be sent over IPC

* LayoutTests/ipc/send-gpu-GetShareableBitmap-RemoteRenderingBackend-expected.txt: Added.
* LayoutTests/ipc/send-gpu-GetShareableBitmap-RemoteRenderingBackend.html: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::backendParameters):
* Source/WebCore/platform/graphics/ImageBuffer.h:
(WebCore::ImageBuffer::create):
(WebCore::ImageBuffer::pixelFormatValidated const):
(WebCore::ImageBuffer::pixelFormat const): Deleted.
* Source/WebCore/platform/graphics/ImageBufferPixelFormat.h: Added.
(WebCore::convertToImageBufferPixelFormat):
(WebCore::convertToPixelFormat):
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
(WebKit::RemoteImageBuffer::getShareableBitmap):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::isSmallLayerBacking):
(WebKit::allocateImageBufferInternal):
(WebKit::RemoteRenderingBackend::createImageBuffer):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::sinkIntoBufferForDifferentThread):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
(WebKit::RemoteImageBufferProxy::create):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::createRemoteImageBuffer):
(WebKit::RemoteRenderingBackendProxy::createDisplayListRecorder):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Tools/TestWebKitAPI/Tests/WebCore/DisplayListRecorderTests.cpp:
(TestWebKitAPI::createTestImageBuffer):

Canonical link: https://commits.webkit.org/279951@main
